azure_agent.go

```diff
@@ -0,0 +1,210 @@
+
+func (a *azureAgent) SendEvent(*EventData) error { return nil }
+
+func (a *azureAgent) SendSpans(spans []Span) error {
+	a.enqueueSpans(spans)
+	return nil
+}
+
+func (a *azureAgent) SendProfiles([]autoprofile.Profile) error { return nil }
+
+func (a *azureAgent) Flush(ctx context.Context) error {
+	a.collectSnapshot()
+
+	if a.snapshot.EntityID == "" {
+		return ErrAgentNotReady
+	}
+
+	from := newServerlessAgentFromS(a.snapshot.EntityID, "azure")
+
+	payload := struct {
+		Metrics metricsPayload `json:"metrics,omitempty"`
+		Spans   []Span         `json:"spans,omitempty"`
+	}{
+		Metrics: metricsPayload{
+			Plugins: []acceptor.PluginPayload{
+				acceptor.NewAzurePluginPayload(a.snapshot.EntityID),
+			},
+		},
+	}
+
+	a.mu.Lock()
+	payload.Spans = make([]Span, len(a.spanQueue))
+	copy(payload.Spans, a.spanQueue)
+	a.spanQueue = a.spanQueue[:0]
+	a.mu.Unlock()
+
+	for i := range payload.Spans {
+		payload.Spans[i].From = from
+	}
+
+	buf := bytes.NewBuffer(nil)
+	if err := json.NewEncoder(buf).Encode(payload); err != nil {
+		return fmt.Errorf("failed to marshal traces payload: %s", err)
+	}
+
+	payloadSize := buf.Len()
+	if payloadSize > maxContentLength {
+		a.logger.Warn(fmt.Sprintf("failed to send the spans. Payload size: %d exceeded max size: %d", payloadSize, maxContentLength))
+		return payloadTooLargeErr
+	}
+
+	req, err := http.NewRequest(http.MethodPost, a.Endpoint+"/bundle", buf)
+	if err != nil {
+		a.enqueueSpans(payload.Spans)
+		return fmt.Errorf("failed to prepare send traces request: %s", err)
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	if err := a.sendRequest(req.WithContext(ctx)); err != nil {
+		a.enqueueSpans(payload.Spans)
+		return fmt.Errorf("failed to send traces, will retry later: %dsec. Error details: %s",
+			flushPeriodInSec, err.Error())
+	}
+
+	return nil
+}
+
+func (a *azureAgent) enqueueSpans(spans []Span) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	a.spanQueue = append(a.spanQueue, spans...)
+}
+
+func (a *azureAgent) sendRequest(req *http.Request) error {
+	req.Header.Set("X-Instana-Host", a.snapshot.Host)
+	req.Header.Set("X-Instana-Key", a.Key)
+
+	resp, err := a.client.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to send request to the serverless agent: %s", err)
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= http.StatusBadRequest {
+		respBody, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			a.logger.Debug("failed to read serverless agent response: ", err.Error())
+			return err
+		}
+
+		a.logger.Info("serverless agent has responded with ", resp.Status, ": ", string(respBody))
+		return err
+	}
+
+	io.CopyN(ioutil.Discard, resp.Body, 1<<20)
+
+	return nil
+}
+
+func (a *azureAgent) collectSnapshot() {
+	if a.snapshot.EntityID != "" {
+		return
+	}
+
+	var subscriptionID, resourceGrp, functionApp string
+	if val, ok := os.LookupEnv("WEBSITE_OWNER_NAME"); ok {
+		arr := strings.Split(val, "+")
+		if len(arr) > 1 {
```

Comment:
Is there any difference if `arr` has more than one element?
My point is, could this be `len(arr) >= 1` instead of `len(arr) > 1` or you have to be sure that the slice has at least 2 elements?